src/Microsoft.Data/ArrowStringColumn.cs

```diff
@@ -0,0 +1,348 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Text;
+using Apache.Arrow;
+using Apache.Arrow.Types;
+
+namespace Microsoft.Data
+{
+    /// <summary>
+    /// An immutable to hold Arrow style strings
+    /// </summary>
+    public partial class ArrowStringColumn : BaseColumn
+    {
+        private IList<DataFrameBuffer<byte>> _dataBuffers;
+        private IList<DataFrameBuffer<int>> _offsetsBuffers;
+        private IList<DataFrameBuffer<byte>> _nullBitMapBuffers;
+
+        private readonly Encoding _encoding = Encoding.UTF8;
+
+        // Need a way to differentiate between columns initialized with default values and those with null values in SetValidityBit
+        internal bool _modifyNullCountWhileIndexing = true;
+
+        public ArrowStringColumn(string name) : base(name, 0, typeof(string))
+        {
+            _dataBuffers = new List<DataFrameBuffer<byte>>();
+            _offsetsBuffers = new List<DataFrameBuffer<int>>();
+            _nullBitMapBuffers = new List<DataFrameBuffer<byte>>();
+        }
+
+        public ArrowStringColumn(string name, DataFrameBuffer<byte> dataBuffer, DataFrameBuffer<int> offsetBuffer, DataFrameBuffer<byte> nullBitMapBuffer, int length) : base(name, length, typeof(string))
+        {
+            if (dataBuffer == null)
+                throw new ArgumentNullException(nameof(dataBuffer));
+            if (offsetBuffer == null)
+                throw new ArgumentNullException(nameof(offsetBuffer));
+            if (nullBitMapBuffer == null)
+                throw new ArgumentNullException(nameof(nullBitMapBuffer));
+            if (length + 1 != offsetBuffer.Length)
+                throw new ArgumentException(nameof(offsetBuffer));
+
+            _dataBuffers = new List<DataFrameBuffer<byte>>();
+            _offsetsBuffers = new List<DataFrameBuffer<int>>();
+            _nullBitMapBuffers = new List<DataFrameBuffer<byte>>();
+
+            _dataBuffers.Add(dataBuffer);
+            _offsetsBuffers.Add(offsetBuffer);
+            _nullBitMapBuffers.Add(nullBitMapBuffer);
+        }
+
+        private long _nullCount;
+        public override long NullCount => _nullCount;
+
+        public bool IsValid(long index) => NullCount == 0 || GetValidityBit(index);
+
+        public bool GetValidityBit(long index)
+        {
+            if ((ulong)index > (ulong)Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
+            // First find the right bitMapBuffer
+            int bitMapIndex = GetBufferIndexContainingRowIndex(ref index);
```

Comment:
I still don't like this pattern of a `ref long` that I know I can cast to an `int` later. I think the method should pass back 2 ints.